### DISCUSSION TITLE
Proposal: next.config.js afterBuild hook

### DISCUSSION THREAD

**Discussion by alvarlagerlof** [vercel/next.js#29820] [Ideas]
I'm making a static og:image generation library that makes use of `.next/server/pages/`. I also expose a wrapper function for `next.config.js` , that provides necessary config for the runtime and cli (soon I hope). 

However, currently only the runtime can read the config, as my main script can't access the next.config.js. The easiest way to run the library would be from a function provided to the by the next.config.js that get's triggered after build. Then I'd also have access to the config everywhere. Currently that is possible for the webpack stage, but webpack seems to only run just _before_ the static pages are generated, and I need to run after. 

I propose a `afterBuild` function that gets called after build. I think this would be very useful for other Next.js libraries too and in turn for the ecosystem.

**Comment by brillout** [comment]
@alvarlagerlof Did you find a workaround?

**Reply by alvarlagerlof** [reply]
I ended up not going for this. It's probably the wrong layer to put yourself it. Easy to make an "extensions" system, but that will likely end up slowing down builds massively. I now think that things should happen in code, or after the build. In between is probably not a good idea. 

In my case, my first solution was to read the .next build output folder for the data I needed. That worked, and resulted in [next-banner](https://www.npmjs.com/package/next-banner). However, I later found out that I might was well just read the generated html files, which lead to [ogimage](https://github.com/alvarlagerlof/ogimage) (unreleased and incomplete), which means it's not even limited to Next.js.

Both of these solutions massively decrease computation time compared to the previous recommended method of running Puppeteer in a serverless function to generate one open graph image per load. This didn't fully work either, since cold boots could be as slow as 8 seconds, causing some messaging services to time out the image, and not display it. I improved on that by using just once instance at build, making it way faster.

However, it was always going to be limited by the website being fully static. I could use one Puppeteer instance only if I only needed to generate the images once. If the data ever changed, a full site build would be needed. This makes it less useful for a lot of sites with dynamic content.

In the end, Vercel probably knew why tuning Puppeteer in serverless functions wasn't a scalable solution at all, which explains. [This thing they introduced](https://vercel.com/blog/introducing-vercel-og-image-generation-fast-dynamic-social-card-images) is probably a better solution in the end, since it can be run dynamically, but without the large overhead of starting ut Puppeteer.

**Reply by brillout** [reply]
Still I think Next.js should make things more customizable (FYI I'm the author https://vite-plugin-ssr.com/ in case you need a more flexible solution).
